torch_geometric/transforms/rooted_subgraphs.py

```diff
@@ -0,0 +1,230 @@
+
+    def __cat_dim__(self, key, value, *args, **kwargs):
+        if 'combined_subgraph' in key:
+            return -1
+        else:
+            return super().__cat_dim__(key, value, *args, **kwargs)
+
+    def rooted_subgraph(self, node_index: int) -> Data:
+        """
+        Returns a :obj:`Data` object representing the rooted subgraph at
+        :obj:`node_index`.
+        """
+        nodes = self.subgraph_nodes_mapper[torch.where(
+            self.subgraph_batch == node_index)]
+        return self.subgraph(nodes)
+
+
+class RootedSubgraph(BaseTransform):
+    r"""
+    Base class for rooted subgraph.
+    The object transforms a Data object to RootedSubgraphData object.
+    """
+    def extract_subgraphs(
+            self, data: Data) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        r""" For input graph with N nodes, extract a rooted subgraph for every
+        node in the graph.
+
+        Returns: a tuple containing
+            subgraph_nodes_mask: N x N dense mask matrix, i-th row indicates
+            the rooted subgraph of node i.
+            hop_indicator (optional): N x N dense matrix, i-th row indicates
+            the distance from other nodes to node i.
+        """
+        raise NotImplementedError
+
+    def __call__(self, data: Data) -> Data:
+        subgraph_nodes_mask, hop_indicator_dense = self.extract_subgraphs(data)
+        subgraph_edges_mask = (subgraph_nodes_mask[:, data.edge_index[0]]
+                               & subgraph_nodes_mask[:, data.edge_index[1]])
+        subgraph_nodes, subgraph_edges, hop_indicator = to_sparse(
+            subgraph_nodes_mask, subgraph_edges_mask, hop_indicator_dense)
+
+        combined_subgraph = combine_subgraph(data.edge_index, subgraph_nodes,
+                                             subgraph_edges,
+                                             num_selected=data.num_nodes,
+                                             num_nodes=data.num_nodes)
+
+        data = RootedSubgraphData(**{k: v for k, v in data})
+        data.subgraph_batch = subgraph_nodes[0]
+        data.subgraph_nodes_mapper = subgraph_nodes[1]
+        data.subgraph_edges_mapper = subgraph_edges[1]
+        data.combined_subgraph = combined_subgraph
+        data.hop_indicator = hop_indicator
+        return data
+
+
+class RootedEgoNets(RootedSubgraph):
+    """ Record rooted k-hop Egonet for each node in the graph.
+    From the `"From Stars to Subgraphs: Uplifting Any GNN with Local Structure
+    Awareness" <https://arxiv.org/pdf/2110.03753.pdf>`_ paper
+
+    Args:
+        num_hops (int): k for k-hop Egonet.
+    """
+    def __init__(self, num_hops: int):
+        super().__init__()
+        self.num_hops = num_hops
+
+    def extract_subgraphs(self,
+                          data: Data) -> Tuple[torch.Tensor, torch.Tensor]:
+        # return k-hop subgraph for all nodes in the graph
+        row, col = data.edge_index
+        sparse_adj = SparseTensor(
```

Comment:
Any reason we do this sparse here? Looks like we are saving dense matrices anyway.

Reply:
I didn't update this yet as I'd like to get rid of the dense matrix if possible